src/midi/midi_alsa.cpp

```diff
@@ -25,19 +25,46 @@
 
 #include <string>
 #include <sstream>
+#include <functional>
 
+#include "logging.h"
+#include "programs.h"
 #include "string_utils.h"
 
 #define ADDR_DELIM ".:"
 
-#if (SND_LIB_MAJOR >= 1) || ((SND_LIB_MAJOR == 0) && (SND_LIB_MINOR >= 6))
-#define snd_seq_flush_output(x)           snd_seq_drain_output(x)
-#define snd_seq_set_client_group(x, name) /* nop */
-#define my_snd_seq_open(seqp)             snd_seq_open(seqp, "hw", SND_SEQ_OPEN_OUTPUT, 0)
-#else
-/* SND_SEQ_OPEN_OUT causes oops on early version of ALSA */
-#define my_snd_seq_open(seqp) snd_seq_open(seqp, SND_SEQ_OPEN)
-#endif
+using port_action_t = std::function<void(snd_seq_client_info_t *client_info,
+                                         snd_seq_port_info_t *port_info)>;
+
+void for_each_alsa_seq_port(port_action_t action)
+{
+	// We can't reuse the sequencer from midi handler, as the function might
+	// be called before that sequencer is created.
+	snd_seq_t *seq;
+	if (snd_seq_open(&seq, "hw", SND_SEQ_OPEN_OUTPUT, 0) != 0) {
```

Comment:
I wonder if using `"hw"` for the device name is too specific.. for example, there are typically multiple hardware items an ALSA configuration, which users (or a higher-level daemon) will rework into various `"default"` sources and sinks.

The [docs](https://www.alsa-project.org/alsa-doc/alsa-lib/group___sequencer.html#ga95462dc59c0319e186cda713ecfb4ed3
) mentions that the sequencer's "name" should usually be "default".

Do you have a working ALSA MIDI configuration you could test "default" with?

Reply:
Good point - string "hw" was used in here only because that was the string used by old API… It seems like ancient versions of ALSA had different port identifiers, which looked like this: `hw:<sequencer>.<port>`, but nowadays they are all `<sequencer>:<port>`. I also found some comments mentioning that "default" is only a modern alias for "hw"…

Anyway - I will change it, we don't care about ancient ALSA versions.